[tanzuobservability exporter] Add consumer for sum metrics.

keep94 opened the pull request:
**Description:** <Describe what has changed.>
Add support for sum metrics.

**Testing:**
Unit tests.

keep94 committed:
[tanzuobservability exporter] Add consumer for sum metrics.

mx-psi commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -249,16 +282,64 @@ func (g *gaugeConsumer) PushInternalMetrics(errs *[]error) {
+	return pdata.MetricDataTypeSum
+}
+
+func (s *sumConsumer) Consume(metric pdata.Metric, errs *[]error) {
```
The specification [gives some indications on how Sums are mapped to a timeseries model](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/metrics/datamodel.md#opentelemetry-protocol-data-model) (PRW-based), in which monotonic Sums are mapped as counters while non-monotonic sums are mapped as gauges. 

What to do here ultimately depends on your timeseries model (which I don't know), so this implementation is not necessarily wrong, but I wanted to bring it up here just in case it helped

keep94 replied:
Thank you for pointing this out. Tanzu observability has a special API call to send a delta sum metric. To send a cumulative sum metric to wavefront, you use the same API call that you would use to send a gauge metric.

keep94 replied:
Thank you for pointing this out. Tanzu observability has a special API call to send a delta sum metric. To send a cumulative sum metric to wavefront, you use the same API call that you would use to send a gauge metric.

oppegard commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -188,6 +190,31 @@ func getValue(numberDataPoint pdata.NumberDataPoint) (float64, error) {
+	errs *[]error,
+	sender gaugeSender,
+	logger *zap.Logger,
+	missingValues *counter) {
```
To make it easier to read where parameters end and the function body starts, please put `) {` on a newline. Same goes for any other methods in this PR that have the same style.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -188,6 +190,31 @@ func getValue(numberDataPoint pdata.NumberDataPoint) (float64, error) {
+// of the metric. Any errors get appended to errs. sender is what sends the
+// gauge metric to tanzu observability. logger is the logger. missingValues
+// keeps track of metrics with missing values.
+func pushGaugeSingleNumberDataPoint(
```
Consider renaming to `pushGaugeNumberDataPoint` – and if we were passing a slice of data points, make it `pushGaugeNumberDataPoints`.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -249,16 +282,64 @@ func (g *gaugeConsumer) PushInternalMetrics(errs *[]error) {
+	}
+}
+
+func (s *sumConsumer) pushSingleNumberDataPoint(
```
For consistency with the Gauge pusher, rename to `pushNumberDataPoint`.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -395,3 +462,136 @@ func copyTags(tags map[string]string) map[string]string {
 	return tagsCopy
 }
+
+func verifySumConsumer(t *testing.T, errorOnSend bool) {
```
This is a large test to comprehend and understand for the three different types of Aggregation Temporality, as well as handling error on send conditions.

Can we break this into 4 individual tests? For example:

- `TestSumConsumerWithDeltaTemporality`
- `TestSumConsumerWithCumulativeTemporality`
- `TestSumConsumerWithUnspecifiedTemporality`
- `TestSumConsumerWhenErrorSending`

keep94 committed:
Function headers to be more readable.

pushGaugeSingleNumberDataPoint->pushGaugeNumberDataPoint
pushSingleNumberDataPoint->pushNumberDataPoint

keep94 committed:
Break verifySumConsumer into four separate tests.

oppegard commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,223 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+			Tags:  map[string]string{"env": "prod"},
+		},
+	}
+	assert.ElementsMatch(t, expected, sender.deltaMetrics)
```
Can we add an assertion like `assert.Empty(t, sender.metrics)`? And add corresponding assertions in the other tests so that both `sender.deltaMetrics` and `sender.metrics` are checked.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,223 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+	assert.Empty(t, errs)
+}
+
+func TestSumConsumerMissingValueNoLogging(t *testing.T) {
```
I think we have pretty good coverage of the desired behavior for Missing Values in  `TestSumConsumerMissingValue`. Consider deleting this test altogether.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,223 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+
+	assert.Len(t, errs, 0)
+	assert.Empty(t, sender.deltaMetrics)
+	require.Len(t, sender.metrics, 1)
+	assert.Equal(t, tobsMetric{
```
Is `require.Len()` necessary if we change the assertion below it to `assert.ElementsMatch()`?

keep94 replied:
Replaced with assert.Contains() which is what I do in the histogram tests.  Later we may publish other internal metrics. What we want to do in this test is make sure that of all the internal metrics pushed, they contain this missing value metric.

oppegard reviewed (changes requested):
Just a couple small changes and then I think we're good!

keep94 committed:
Minor changes to tests.

oppegard reviewed (approved):
Looks good!
